<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Model fitting &#8212; imGLAD 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="top" title="imGLAD 1.0 documentation" href="index.html" />
    <link rel="next" title="Probability estimation" href="probEstimate.html" />
    <link rel="prev" title="Getting started" href="introduction.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="probEstimate.html" title="Probability estimation"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="introduction.html" title="Getting started"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">imGLAD 1.0 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="model-fitting">
<h1>Model fitting<a class="headerlink" href="#model-fitting" title="Permalink to this headline">¶</a></h1>
<p>imGLAD attempts to solve a binary classification problem. The positive examples in this case are all the samples were the target genome is present. The negative examples are all the samples were the negative genome is absent. <code class="docutils literal"><span class="pre">fitModel.py</span></code> generates a series of metagenomic datasets, 50% of this metagenomic datasets will contain reads from the target genome, these are labeled as positive samples.  The remaining 50% will <strong>not</strong> contain any reads from the target genome. These sets of samples are used to separate positive from negative samples using a logistic model.</p>
<div class="section" id="general-options">
<h2>General options<a class="headerlink" href="#general-options" title="Permalink to this headline">¶</a></h2>
<p>For a target genome a logistic model is built. The target genome is provided in a Fasta file. The associated options are.</p>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">-t</span></kbd></td>
<td>The target genome to be detected in FASTA format.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-s<var>p</var></span></kbd></td>
<td>The species of the genome written as binomial name in quotations. (e.g. &#8220;Escherichia coli&#8221;)</td></tr>
</tbody>
</table>
<p>The option <code class="docutils literal"><span class="pre">-t</span></code> refers to the file of the target genome. The option <code class="docutils literal"><span class="pre">-sp</span></code> refers to the species name, this field is used to avoid the inclusion of closely related gnoes in the training set (see below). If the species name is unknown this field can be left blank <code class="docutils literal"><span class="pre">''</span></code>.</p>
</div>
<div class="section" id="the-training-dataset">
<h2>The training dataset<a class="headerlink" href="#the-training-dataset" title="Permalink to this headline">¶</a></h2>
<p>Metagenomic datasets are generated by random generation of reads from the selected genomes. The associated options are.</p>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">-l <var>GENOMES</var></span></kbd></td>
<td>The genomes list to create the training dataset</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-s <var>TRAIN_SIZE</var></span></kbd></td>
<td>Number of genomes included in the training dataset (Incompatible with -l). (default : 200)</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-e <var>TRAINING_EXAMPLES</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Number of training examples (metagenomic datasets) used to train the model by default 200 (100 positive and 100 negative examples. (default : 100)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-j <var>PLATFORM</var></span></kbd></td>
<td>The sequencing platform used to generate the reads in the datasets. (default : illumina)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-r <var>NUM_READS</var></span></kbd></td>
<td>Number of reads per training example (Metagenomic dataset simulated). (default : 1000000)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-d <var>READ_LENGTH</var></span></kbd></td>
<td>Average read length for the simulated datasets. (default : 150)</td></tr>
</tbody>
</table>
</div>
<div class="section" id="using-your-own-list-of-genomes">
<h2>Using your own list of genomes<a class="headerlink" href="#using-your-own-list-of-genomes" title="Permalink to this headline">¶</a></h2>
<p>If you need a simulated metagenome with a specific set of genomes from NCBI you can provide imGLAD with a specific list of the accesion numbers for the genomes. imGLAD will download the genomes in the list and create metagenomic datasets with only these genomes in order to train the model. The target model needs to be provided in a separate file.</p>
<div class="section" id="example-1">
<h3>Example 1<a class="headerlink" href="#example-1" title="Permalink to this headline">¶</a></h3>
<p>A genome list could be the following.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">GCA_000148745</span><span class="o">.</span><span class="mi">1</span>
<span class="n">GCA_000521945</span><span class="o">.</span><span class="mi">1</span>
<span class="n">GCA_000492815</span><span class="o">.</span><span class="mi">1</span>
<span class="n">GCA_001444575</span><span class="o">.</span><span class="mi">1</span>
<span class="n">GCA_000007845</span><span class="o">.</span><span class="mi">1</span>
</pre></div>
</div>
<p>Which will include genomes from:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Pseudomonas</span> <span class="n">aeruginosa</span> <span class="mi">39016</span>
<span class="n">Citrobacter</span> <span class="n">freundii</span> <span class="n">UCI</span> <span class="mi">32</span>
<span class="n">Klebsiella</span> <span class="n">oxytoca</span> <span class="n">MGH</span> <span class="mi">42</span>
<span class="n">Clostridium</span> <span class="n">sporogenes</span>
<span class="n">Bacillus</span> <span class="n">anthracis</span> <span class="nb">str</span><span class="o">.</span> <span class="n">Ames</span>
</pre></div>
</div>
<p>A simple run of <code class="docutils literal"><span class="pre">fitModel.py</span></code> could be:</p>
<p><code class="docutils literal"><span class="pre">$&gt;./fitModel.py</span> <span class="pre">-t</span> <span class="pre">genome.fa</span> <span class="pre">-sp</span> <span class="pre">'Escherichia</span> <span class="pre">coli'</span> <span class="pre">-l</span> <span class="pre">genomes.txt</span></code></p>
<p>This will create a model for <em>Escherichia coli</em>, from the target <code class="docutils literal"><span class="pre">genome.fa</span></code> and the genome list <code class="docutils literal"><span class="pre">genomes.txt</span></code>.</p>
</div>
</div>
<div class="section" id="using-random-genomes-from-ncbi">
<h2>Using random genomes from NCBI<a class="headerlink" href="#using-random-genomes-from-ncbi" title="Permalink to this headline">¶</a></h2>
<p>If you don&#8217;t have any preference regarding the genomes used to build the training set, imGLAD can built a model with a random number of genomes from NCBI, the user can prive the number of genomes to be used.</p>
<div class="section" id="example-2">
<h3>Example 2<a class="headerlink" href="#example-2" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">$&gt;./fitModel.py</span> <span class="pre">-t</span> <span class="pre">genome.fa</span> <span class="pre">-sp</span> <span class="pre">'Escherichia</span> <span class="pre">coli'</span> <span class="pre">-s</span> <span class="pre">100</span></code></p>
<p>This will create a model for <em>Escherichia coli</em>, from the target <code class="docutils literal"><span class="pre">genome.fa</span></code> using 100 randomly selected genomes from NCBI as background metagenomic signal.</p>
</div>
</div>
<div class="section" id="output-files">
<h2>Output files<a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">parameters</span><span class="o">.</span><span class="n">txt</span>
<span class="n">trainingGenomes</span><span class="o">.</span><span class="n">txt</span>
<span class="n">detectionLimit</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>After running <code class="docutils literal"><span class="pre">fitModel.py</span></code> 3 files will be created in your working directory. a <code class="docutils literal"><span class="pre">parameters.txt</span></code> file will contain the values for the regresssion parameters of the model, this file should be provided to <code class="docutils literal"><span class="pre">probEstimate.py</span></code> as an input under the option <code class="docutils literal"><span class="pre">-p</span></code>. Additionally a file <code class="docutils literal"><span class="pre">trainingGenomes.txt</span></code> is also created, this file will contain a list of the files used to create the training set. This should give you an idea of the complexity and diversity under which the model was created. Finally a <code class="docutils literal"><span class="pre">detectionLimit.txt</span></code> file, this file will include a calculation for the detection limit with a 95% confidence.</p>
</div>
<div class="section" id="general-use-examples">
<h2>General use examples<a class="headerlink" href="#general-use-examples" title="Permalink to this headline">¶</a></h2>
<div class="section" id="example-3">
<h3>Example 3<a class="headerlink" href="#example-3" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">$&gt;./fitModel.py</span> <span class="pre">-t</span> <span class="pre">genome.fa</span> <span class="pre">-sp</span> <span class="pre">'Escherichia</span> <span class="pre">coli'</span> <span class="pre">-e</span> <span class="pre">100</span></code></p>
<p>Specifying <code class="docutils literal"><span class="pre">-e</span></code> instructs <code class="docutils literal"><span class="pre">fitModel.py</span></code> to create a given number of training metagenomic datasets, the value specified with <code class="docutils literal"><span class="pre">-e</span></code> represents the number of positive and negative metagenomic datasets (with and withoureads from the target genome respectively). The default is to create 100 of each.</p>
<p><strong>NOTE:</strong> A small numnber of training examples will take less computational resources (disk space and running time) but will likely yield inaccurate results. Perfect separation errors can occur as a consequence of small values of <code class="docutils literal"><span class="pre">-e</span></code>. This means that the positive and negative datasets are too far apart for imGLAD to establish a detection limit. If you get this error you should increase the training dataset size by increasing the value of <code class="docutils literal"><span class="pre">-e</span></code></p>
</div>
<div class="section" id="example-4">
<h3>Example 4<a class="headerlink" href="#example-4" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">$&gt;./fitModel.py</span> <span class="pre">-t</span> <span class="pre">genome.fa</span> <span class="pre">-sp</span> <span class="pre">'Escherichia</span> <span class="pre">coli'</span> <span class="pre">-j</span> <span class="pre">illumina</span></code></p>
<p>Specifying <code class="docutils literal"><span class="pre">-j</span></code> instructs <code class="docutils literal"><span class="pre">fitModel.py</span></code> to use a specific platform for read generation in this case Illumina. Altough solid and Roche 454 technologies are also available.</p>
</div>
<div class="section" id="example-5">
<h3>Example 5<a class="headerlink" href="#example-5" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">$&gt;./fitModel.py</span> <span class="pre">-t</span> <span class="pre">genome.fa</span> <span class="pre">-sp</span> <span class="pre">'Escherichia</span> <span class="pre">coli'</span> <span class="pre">-r</span> <span class="pre">1000000</span></code></p>
<p>Specifying <code class="docutils literal"><span class="pre">-r</span></code> instructs <code class="docutils literal"><span class="pre">fitModel.py</span></code> to generate a particular number of reads in each training example. The  default is 1 millon.</p>
</div>
<div class="section" id="example-6">
<h3>Example 6<a class="headerlink" href="#example-6" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">$&gt;./fitModel.py</span> <span class="pre">-t</span> <span class="pre">genome.fa</span> <span class="pre">-sp</span> <span class="pre">'Escherichia</span> <span class="pre">coli'</span> <span class="pre">-d</span> <span class="pre">150</span></code></p>
<p>Specifying <code class="docutils literal"><span class="pre">-d</span></code> designates the length of the reads in the training datasets.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Model fitting</a><ul>
<li><a class="reference internal" href="#general-options">General options</a></li>
<li><a class="reference internal" href="#the-training-dataset">The training dataset</a></li>
<li><a class="reference internal" href="#using-your-own-list-of-genomes">Using your own list of genomes</a><ul>
<li><a class="reference internal" href="#example-1">Example 1</a></li>
</ul>
</li>
<li><a class="reference internal" href="#using-random-genomes-from-ncbi">Using random genomes from NCBI</a><ul>
<li><a class="reference internal" href="#example-2">Example 2</a></li>
</ul>
</li>
<li><a class="reference internal" href="#output-files">Output files</a></li>
<li><a class="reference internal" href="#general-use-examples">General use examples</a><ul>
<li><a class="reference internal" href="#example-3">Example 3</a></li>
<li><a class="reference internal" href="#example-4">Example 4</a></li>
<li><a class="reference internal" href="#example-5">Example 5</a></li>
<li><a class="reference internal" href="#example-6">Example 6</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="introduction.html"
                        title="previous chapter">Getting started</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="probEstimate.html"
                        title="next chapter">Probability estimation</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/fitModel.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="probEstimate.html" title="Probability estimation"
             >next</a> |</li>
        <li class="right" >
          <a href="introduction.html" title="Getting started"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">imGLAD 1.0 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Juan C. Castro.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.9.
    </div>
  </body>
</html>