<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Probability estimation &#8212; imGLAD 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="top" title="imGLAD 1.0 documentation" href="index.html" />
    <link rel="prev" title="Model fitting" href="fitModel.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="fitModel.html" title="Model fitting"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">imGLAD 1.0 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="probability-estimation">
<h1>Probability estimation<a class="headerlink" href="#probability-estimation" title="Permalink to this headline">¶</a></h1>
<p><code class="docutils literal"><span class="pre">probEstimate.py</span></code> estimates the likelihood of presence of your target genome in a metagenomic sample (one or more).</p>
<div class="section" id="getting-started">
<h2>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this headline">¶</a></h2>
<p>First you need to run <code class="docutils literal"><span class="pre">fitModel.py</span></code> and get the parameters file it generatesi (<code class="docutils literal"><span class="pre">parameters.txt</span></code>).</p>
<p>Before running <code class="docutils literal"><span class="pre">probEstimate.py</span></code> you should also run a Blastn or Blat of your metagenomic sample against the target genome you created the model with. The output should be stored in tabular format.</p>
</div>
<div class="section" id="general-options">
<h2>General options<a class="headerlink" href="#general-options" title="Permalink to this headline">¶</a></h2>
<div class="section" id="required-arguments">
<h3>Required arguments<a class="headerlink" href="#required-arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">-t <var>TARGET</var></span></kbd></td>
<td>Subject sequences (ref) in FastA format.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-m <var>MAP</var></span></kbd></td>
<td>One or more Tabular BLAST files of reads vs genes (or contigs).</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-p <var>PARAM</var></span></kbd></td>
<td>Parameters file obtained from fitModel.py</td></tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="single-mode">
<h2>Single mode<a class="headerlink" href="#single-mode" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal"><span class="pre">parameters.txt</span></code> file will save parameters for a model based on sequencing breath only and a model based o sequeincing breadth and depth. Single mode will use sequencing breadth parameters to estimate the likelihood of presence.</p>
<div class="section" id="example-1">
<h3>Example 1<a class="headerlink" href="#example-1" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">$&gt;./probEstimate.py</span> <span class="pre">-t</span> <span class="pre">genome.fa</span> <span class="pre">-m</span> <span class="pre">blastTab_01.tbl</span> <span class="pre">-p</span> <span class="pre">parameters.txt</span> <span class="pre">-l</span> <span class="pre">single</span></code></p>
<p>The above line will estimate the likelihood of presence of the target genome in the metagenomic sample, based on sequencing breadth, the results will be displayed in the screen like.</p>
<table border="1" class="docutils">
<colgroup>
<col width="20%" />
<col width="35%" />
<col width="31%" />
<col width="14%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>Filename</td>
<td>Sequencing breadth</td>
<td>Sequencing depth</td>
<td>P-value</td>
</tr>
<tr class="row-even"><td>file01.tsv</td>
<td>float</td>
<td>float</td>
<td>float</td>
</tr>
</tbody>
</table>
<p>You can save these results by redirecting the output to a file.</p>
<p><code class="docutils literal"><span class="pre">$&gt;./probEstimate.py</span> <span class="pre">-t</span> <span class="pre">genome.fa</span> <span class="pre">-m</span> <span class="pre">blastTab_01.tbl</span> <span class="pre">-p</span> <span class="pre">parameters.txt</span> <span class="pre">-l</span> <span class="pre">single</span>&#160; <span class="pre">&gt;</span> <span class="pre">results.tsv</span></code></p>
</div>
</div>
<div class="section" id="general-mode">
<h2>General mode<a class="headerlink" href="#general-mode" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">fitModel.py</span></code> estimates parameters for the model based on sequencing depth as well. If general mode is used both variables will be used to calculate the precense probability.</p>
<div class="section" id="example-2">
<h3>Example 2<a class="headerlink" href="#example-2" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">$&gt;./probEstimate.py</span> <span class="pre">-t</span> <span class="pre">genome.fa</span> <span class="pre">-m</span> <span class="pre">blastTab_01.tbl</span> <span class="pre">-p</span> <span class="pre">parameters.txt</span> <span class="pre">-l</span> <span class="pre">general</span></code></p>
<p>Will instruct <code class="docutils literal"><span class="pre">probEstimate.py</span></code> to include sequencing depth in the calculation of likelihood for presence.</p>
<p><strong>NOTE:</strong> Using general mode can lead to false positive identification due to high mapping of conserved regions, plasmids, or mobile elements.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Probability estimation</a><ul>
<li><a class="reference internal" href="#getting-started">Getting started</a></li>
<li><a class="reference internal" href="#general-options">General options</a><ul>
<li><a class="reference internal" href="#required-arguments">Required arguments</a></li>
</ul>
</li>
<li><a class="reference internal" href="#single-mode">Single mode</a><ul>
<li><a class="reference internal" href="#example-1">Example 1</a></li>
</ul>
</li>
<li><a class="reference internal" href="#general-mode">General mode</a><ul>
<li><a class="reference internal" href="#example-2">Example 2</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="fitModel.html"
                        title="previous chapter">Model fitting</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/probEstimate.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="fitModel.html" title="Model fitting"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">imGLAD 1.0 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Juan C. Castro.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.9.
    </div>
  </body>
</html>