% Generated by Sphinx.
\def\sphinxdocclass{report}
\newif\ifsphinxKeepOldNames \sphinxKeepOldNamestrue
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\usepackage{iftex}

\ifPDFTeX
  \usepackage[utf8]{inputenc}
\fi
\ifdefined\DeclareUnicodeCharacter
  \DeclareUnicodeCharacter{00A0}{\nobreakspace}
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{longtable}
\usepackage{sphinx}
\usepackage{multirow}
\usepackage{eqparbox}


\addto\captionsenglish{\renewcommand{\figurename}{Fig.\@ }}
\addto\captionsenglish{\renewcommand{\tablename}{Table }}
\SetupFloatingEnvironment{literal-block}{name=Listing }

\addto\extrasenglish{\def\pageautorefname{page}}

\setcounter{tocdepth}{1}


\title{imGLAD Documentation}
\date{Mar 10, 2017}
\release{1.0}
\author{Juan C. Castro}
\newcommand{\sphinxlogo}{}
\renewcommand{\releasename}{Release}
\makeindex

\makeatletter
\def\PYG@reset{\let\PYG@it=\relax \let\PYG@bf=\relax%
    \let\PYG@ul=\relax \let\PYG@tc=\relax%
    \let\PYG@bc=\relax \let\PYG@ff=\relax}
\def\PYG@tok#1{\csname PYG@tok@#1\endcsname}
\def\PYG@toks#1+{\ifx\relax#1\empty\else%
    \PYG@tok{#1}\expandafter\PYG@toks\fi}
\def\PYG@do#1{\PYG@bc{\PYG@tc{\PYG@ul{%
    \PYG@it{\PYG@bf{\PYG@ff{#1}}}}}}}
\def\PYG#1#2{\PYG@reset\PYG@toks#1+\relax+\PYG@do{#2}}

\expandafter\def\csname PYG@tok@gd\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gu\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@gt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PYG@tok@gs\endcsname{\let\PYG@bf=\textbf}
\expandafter\def\csname PYG@tok@gr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@cm\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@vg\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@vi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@mh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@cs\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\colorbox[rgb]{1.00,0.94,0.94}{\strut ##1}}}
\expandafter\def\csname PYG@tok@ge\endcsname{\let\PYG@it=\textit}
\expandafter\def\csname PYG@tok@vc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@il\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@go\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.20,0.20,0.20}{##1}}}
\expandafter\def\csname PYG@tok@cp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@gi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gh\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@ni\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.84,0.33,0.22}{##1}}}
\expandafter\def\csname PYG@tok@nl\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.13,0.44}{##1}}}
\expandafter\def\csname PYG@tok@nn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@no\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.38,0.68,0.84}{##1}}}
\expandafter\def\csname PYG@tok@na\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@nb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@nd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.33,0.33,0.33}{##1}}}
\expandafter\def\csname PYG@tok@ne\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.49}{##1}}}
\expandafter\def\csname PYG@tok@si\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.44,0.63,0.82}{##1}}}
\expandafter\def\csname PYG@tok@s2\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@nt\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.45}{##1}}}
\expandafter\def\csname PYG@tok@nv\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@s1\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@ch\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@m\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@gp\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@sh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@ow\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@sx\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@bp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c1\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@o\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PYG@tok@kc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@mf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@err\endcsname{\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PYG@tok@mb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@ss\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.32,0.47,0.09}{##1}}}
\expandafter\def\csname PYG@tok@sr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.14,0.33,0.53}{##1}}}
\expandafter\def\csname PYG@tok@mo\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@mi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@cpf\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@kr\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@s\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@kp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@w\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PYG@tok@kt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.56,0.13,0.00}{##1}}}
\expandafter\def\csname PYG@tok@sc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@k\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@se\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sd\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}

\def\PYGZbs{\char`\\}
\def\PYGZus{\char`\_}
\def\PYGZob{\char`\{}
\def\PYGZcb{\char`\}}
\def\PYGZca{\char`\^}
\def\PYGZam{\char`\&}
\def\PYGZlt{\char`\<}
\def\PYGZgt{\char`\>}
\def\PYGZsh{\char`\#}
\def\PYGZpc{\char`\%}
\def\PYGZdl{\char`\$}
\def\PYGZhy{\char`\-}
\def\PYGZsq{\char`\'}
\def\PYGZdq{\char`\"}
\def\PYGZti{\char`\~}
% for compatibility with earlier versions
\def\PYGZat{@}
\def\PYGZlb{[}
\def\PYGZrb{]}
\makeatother

\renewcommand\PYGZsq{\textquotesingle}

\begin{document}

\maketitle
\tableofcontents
\phantomsection\label{index::doc}



\chapter{Table of Contents:}
\label{index:imglad-bacterial-genome-detection-in-metagenomic-datasets}\label{index:table-of-contents}

\section{Getting started}
\label{introduction:getting-started}\label{introduction::doc}

\subsection{What is imGLAD?}
\label{introduction:what-is-imglad}
imGLAD is a computational tool for detection of bacterial genomes in metagenomic datasets. imGLAD has two steps fitting a model and predicting data. The fitting step simulates metagenomic datasets and randoly spikes them with the target genome. This is used to create a logistic model to discriminate positive samples from negative samples. The prediction step uses the model paramters to estimate the likelihood of presence of the target genome in a given metagenomic sample.


\subsection{Obtaining imGLAD}
\label{introduction:obtaining-imglad}
Clone the git repository
\begin{quote}

\sphinxcode{\$\textgreater{} git clone https://github.com/jccastrog/imGLAD}
\end{quote}

You can also download the zip file from the GitHub site \url{https://github.com/jccastrog/imGLAD}


\subsection{System requirements}
\label{introduction:system-requirements}
Python 2.7 or higher.

ART 2.5.8 or higher.

Either BLAST 2.2.28 (or higher) or BLAT (any version)


\subsubsection{Python requirements}
\label{introduction:python-requirements}
numpy, scipy, biopython, gzip, screed, statsmodel (optional)


\section{Model fitting}
\label{fitModel:model-fitting}\label{fitModel::doc}
imGLAD attempts to solve a binary classification problem. The positive examples in this case are all the samples were the target genome is present. The negative examples are all the samples were the negative genome is absent. \sphinxcode{fitModel.py} generates a series of metagenomic datasets, 50\% of this metagenomic datasets will contain reads from the target genome, these are labeled as positive samples.  The remaining 50\% will \textbf{not} contain any reads from the target genome. These sets of samples are used to separate positive from negative samples using a logistic model.


\subsection{General options}
\label{fitModel:general-options}
For a target genome a logistic model is built. The target genome is provided in a Fasta file. The associated options are.
\begin{optionlist}{3cm}
\item [-t]  
The target genome to be detected in FASTA format.
\item [-sp]  
The species of the genome written as binomial name in quotations. (e.g. ``Escherichia coli'')
\end{optionlist}

The option \sphinxcode{-t} refers to the file of the target genome. The option \sphinxcode{-sp} refers to the species name, this field is used to avoid the inclusion of closely related gnoes in the training set (see below). If the species name is unknown this field can be left blank \sphinxcode{'{'}}.


\subsection{The training dataset}
\label{fitModel:the-training-dataset}
Metagenomic datasets are generated by random generation of reads from the selected genomes. The associated options are.
\begin{optionlist}{3cm}
\item [-l GENOMES]  
The genomes list to create the training dataset
\item [-s TRAIN\_SIZE]  
Number of genomes included in the training dataset (Incompatible with -l). (default : 200)
\item [-e TRAINING\_EXAMPLES]  
Number of training examples (metagenomic datasets) used to train the model by default 200 (100 positive and 100 negative examples. (default : 100)
\item [-j PLATFORM]  
The sequencing platform used to generate reads in the datasets. (default : illumina)
\item [-r NUM\_READS]  
Number of reads per training example (Metagenomic dataset simulated). (default : 1000000)
\item [-d READ\_LENGTH]  
Average read length for the simulated datasets. (default : 150)
\end{optionlist}


\subsection{Using your own list of genomes}
\label{fitModel:using-your-own-list-of-genomes}
If you need a simulated metagenome with a specific set of genomes from NCBI you can provide imGLAD with a specific list of the accesion numbers for the genomes. imGLAD will download the genomes in the list and create metagenomic datasets with only these genomes in order to train the model. The target model needs to be provided in a separate file.


\subsubsection{Example 1}
\label{fitModel:example-1}
A genome list could be the following.

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{GCA\PYGZus{}000148745}\PYG{o}{.}\PYG{l+m+mi}{1}
\PYG{n}{GCA\PYGZus{}000521945}\PYG{o}{.}\PYG{l+m+mi}{1}
\PYG{n}{GCA\PYGZus{}000492815}\PYG{o}{.}\PYG{l+m+mi}{1}
\PYG{n}{GCA\PYGZus{}001444575}\PYG{o}{.}\PYG{l+m+mi}{1}
\PYG{n}{GCA\PYGZus{}000007845}\PYG{o}{.}\PYG{l+m+mi}{1}
\end{Verbatim}

Which will include genomes from:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Pseudomonas} \PYG{n}{aeruginosa} \PYG{l+m+mi}{39016}
\PYG{n}{Citrobacter} \PYG{n}{freundii} \PYG{n}{UCI} \PYG{l+m+mi}{32}
\PYG{n}{Klebsiella} \PYG{n}{oxytoca} \PYG{n}{MGH} \PYG{l+m+mi}{42}
\PYG{n}{Clostridium} \PYG{n}{sporogenes}
\PYG{n}{Bacillus} \PYG{n}{anthracis} \PYG{n+nb}{str}\PYG{o}{.} \PYG{n}{Ames}
\end{Verbatim}

A simple run of \sphinxcode{fitModel.py} could be:

\sphinxcode{\$\textgreater{}./fitModel.py -t genome.fa -sp 'Escherichia coli' -l genomes.txt}

This will create a model for \emph{Escherichia coli}, from the target \sphinxcode{genome.fa} and the genome list \sphinxcode{genomes.txt}.


\subsection{Using random genomes from NCBI}
\label{fitModel:using-random-genomes-from-ncbi}
If you don't have any preference regarding the genomes used to build the training set, imGLAD can built a model with a random number of genomes from NCBI. The user can provide the number of genomes to be used.


\subsubsection{Example 2}
\label{fitModel:example-2}
\sphinxcode{\$\textgreater{}./fitModel.py -t genome.fa -sp 'Escherichia coli' -s 100}

This will create a model for \emph{Escherichia coli}, from the target \sphinxcode{genome.fa} using 100 randomly selected genomes from NCBI as background metagenomic signal.


\subsection{Output files}
\label{fitModel:output-files}
\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{parameters}\PYG{o}{.}\PYG{n}{txt}
\PYG{n}{trainingGenomes}\PYG{o}{.}\PYG{n}{txt}
\PYG{n}{detectionLimit}\PYG{o}{.}\PYG{n}{txt}
\end{Verbatim}

After running \sphinxcode{fitModel.py} 3 files will be created in your working directory. a \sphinxcode{parameters.txt} file will contain the values for the regresssion parameters of the model, this file should be provided to \sphinxcode{probEstimate.py} as an input under the option \sphinxcode{-p}. Additionally a file \sphinxcode{trainingGenomes.txt} is also created, this file will contain a list of the files used to create the training set. This should give you an idea of the complexity and diversity under which the model was created. Finally a \sphinxcode{detectionLimit.txt} file, this file will include a calculation for the detection limit with a 95\% confidence.


\subsection{General use examples}
\label{fitModel:general-use-examples}

\subsubsection{Example 3}
\label{fitModel:example-3}
\sphinxcode{\$\textgreater{}./fitModel.py -t genome.fa -sp 'Escherichia coli' -e 100}

Specifying \sphinxcode{-e} instructs \sphinxcode{fitModel.py} to create a given number of training metagenomic datasets, the value specified with \sphinxcode{-e} represents the number of positive and negative metagenomic datasets (with and withoureads from the target genome respectively). The default is to create 100 of each.

\textbf{NOTE:} A small numnber of training examples will take less computational resources (disk space and running time) but will likely yield inaccurate results. Perfect separation errors can occur as a consequence of small values of \sphinxcode{-e}. This means that the positive and negative datasets are too far apart for imGLAD to establish a detection limit. If you get this error you should increase the training dataset size by increasing the value of \sphinxcode{-e}


\subsubsection{Example 4}
\label{fitModel:example-4}
\sphinxcode{\$\textgreater{}./fitModel.py -t genome.fa -sp 'Escherichia coli' -j illumina}

Specifying \sphinxcode{-j} instructs \sphinxcode{fitModel.py} to use a specific platform for read generation in this case Illumina. Although solid and Roche 454 technologies are also available.


\subsubsection{Example 5}
\label{fitModel:example-5}
\sphinxcode{\$\textgreater{}./fitModel.py -t genome.fa -sp 'Escherichia coli' -r 1000000}

Specifying \sphinxcode{-r} instructs \sphinxcode{fitModel.py} to generate a particular number of reads in each training example. The  default is 1 millon.


\subsubsection{Example 6}
\label{fitModel:example-6}
\sphinxcode{\$\textgreater{}./fitModel.py -t genome.fa -sp 'Escherichia coli' -d 150}

Specifying \sphinxcode{-d} designates the length of the reads in the training datasets.


\section{Probability estimation}
\label{probEstimate:probability-estimation}\label{probEstimate::doc}
\sphinxcode{probEstimate.py} estimates the likelihood of presence of your target genome in a metagenomic sample (one or more).


\subsection{Getting started}
\label{probEstimate:getting-started}
First you need to run \sphinxcode{fitModel.py} and get the parameters file it generatesi (\sphinxcode{parameters.txt}).

Before running \sphinxcode{probEstimate.py} you should also run a Blastn or Blat of your metagenomic sample against the target genome you created the model with. The output should be stored in tabular format.


\subsection{General options}
\label{probEstimate:general-options}

\subsubsection{Required arguments}
\label{probEstimate:required-arguments}\begin{optionlist}{3cm}
\item [-t TARGET]  
Subject sequences (ref) in FastA format.
\item [-m MAP]  
One or more Tabular BLAST files of reads vs genes (or contigs).
\item [-p PARAM]  
Parameters file obtained from fitModel.py
\end{optionlist}


\subsection{Single mode}
\label{probEstimate:single-mode}
The \sphinxcode{parameters.txt} file will save parameters for a model based on sequencing breath only and a model based o sequeincing breadth and depth. Single mode will use sequencing breadth parameters to estimate the likelihood of presence.


\subsubsection{Example 1}
\label{probEstimate:example-1}
\sphinxcode{\$\textgreater{}./probEstimate.py -t genome.fa -m blastTab\_01.tbl -p parameters.txt -l single}

The above line will estimate the likelihood of presence of the target genome in the metagenomic sample, based on sequencing breadth, the results will be displayed in the screen like.

\noindent\begin{tabulary}{\linewidth}{|L|L|L|L|}
\hline

Filename
&
Sequencing breadth
&
Sequencing depth
&
P-value
\\
\hline
file01.tsv
&
float
&
float
&
float
\\
\hline\end{tabulary}


You can save these results by redirecting the output to a file.

\sphinxcode{\$\textgreater{}./probEstimate.py -t genome.fa -m blastTab\_01.tbl -p parameters.txt -l single  \textgreater{} results.tsv}


\subsection{General mode}
\label{probEstimate:general-mode}
\sphinxcode{fitModel.py} estimates parameters for the model based on sequencing depth as well. If general mode is used both variables will be used to calculate the precense probability.


\subsubsection{Example 2}
\label{probEstimate:example-2}
\sphinxcode{\$\textgreater{}./probEstimate.py -t genome.fa -m blastTab\_01.tbl -p parameters.txt -l general}

Will instruct \sphinxcode{probEstimate.py} to include sequencing depth in the calculation of likelihood for presence.

\textbf{NOTE:} Using general mode can lead to false positive identification due to high mapping of conserved regions, plasmids, or mobile elements.



\renewcommand{\indexname}{Index}
\printindex
\end{document}
